### PR TITLE
[SPARK-21276][CORE] Update lz4-java to the latest (v1.4.0)

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr updated `lz4-java` to the latest (v1.4.0) and removed custom `LZ4BlockInputStream`. We currently use custom `LZ4BlockInputStream` to read concatenated byte stream in shuffle. But, this functionality has been implemented in the latest lz4-java (https://github.com/lz4/lz4-java/pull/105). So, we might update the latest to remove the custom `LZ4BlockInputStream`.

Major diffs between the latest release and v1.3.0 in the master are as follows (https://github.com/lz4/lz4-java/compare/62f7547abb0819d1ca1e669645ee1a9d26cd60b0...6d4693f56253fcddfad7b441bb8d917b182efa2d);
- fixed NPE in XXHashFactory similarly
- Don't place resources in default package to support shading
- Fixes ByteBuffer methods failing to apply arrayOffset() for array-backed
- Try to load lz4-java from java.library.path, then fallback to bundled
- Add ppc64le binary
- Add s390x JNI binding
- Add basic LZ4 Frame v1.5.0 support
- enable aarch64 support for lz4-java
- Allow unsafeInstance() for ppc64le archiecture
- Add unsafeInstance support for AArch64
- Support 64-bit JNI build on Solaris
- Avoid over-allocating a buffer
- Allow EndMark to be incompressible for LZ4FrameInputStream.
- Concat byte stream

## How was this patch tested?
Existing tests.

### CODE REVIEW

**Review by srowen — APPROVED** [review]

File: core/src/main/java/org/apache/spark/io/LZ4BlockInputStream.java [hunk 1]
```diff
@@ -1,260 +0,0 @@
...
- * support {@link #mark(int)}/{@link #reset()}.
- * @see net.jpountz.lz4.LZ4BlockOutputStream
- *
- * This is based on net.jpountz.lz4.LZ4BlockInputStream
- *
- * changes: https://github.com/davies/lz4-java/commit/cc1fa940ac57cc66a0b937300f805d37e2bf8411
- *
- * TODO: merge this into upstream
- */
-public final class LZ4BlockInputStream extends FilterInputStream {
```
Comment (srowen) [inline core/src/main/java/org/apache/spark/io/LZ4BlockInputStream.java]: Yeah I guess this needs a MiMa exclude. It's technically public but nobody should have ever referenced this directly. The codec class is separate.

**Comment by maropu** [comment]
I'll update soon.

File: core/src/main/java/org/apache/spark/io/LZ4BlockInputStream.java (hunk 1 above)
Comment (maropu) [inline core/src/main/java/org/apache/spark/io/LZ4BlockInputStream.java]: yea, thanks! I'll check and update soon.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by maropu** [comment]
@dongjoon-hyun Thanks for your test re-triggers!

**Comment by dongjoon-hyun** [comment]
It's because I like your PR. :)

**Comment by kiszk** [comment]
LGTM

**Comment by srowen** [comment]
Merged to master

File: project/MimaExcludes.scala [hunk 2]
```diff
@@ -41,7 +41,10 @@ object MimaExcludes {
 
     // [SPARK-19937] Add remote bytes read to disk.
     ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.status.api.v1.ShuffleReadMetrics.this"),
-    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.status.api.v1.ShuffleReadMetricDistributions.this")
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.status.api.v1.ShuffleReadMetricDistributions.this"),
+
+    // [SPARK-21276] Update lz4-java to the latest (v1.4.0)
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.io.LZ4BlockInputStream")
```
Comment (zsxwing) [inline project/MimaExcludes.scala]: @srowen This is a breaking change. We should not remove a public class that is in the api docs: http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.io.LZ4BlockInputStream

Comment (srowen) [reply project/MimaExcludes.scala]: It's "public" only insofar as it has to be in Java to use it this way. There's no case where a user should or would use this class directly.

Comment (zsxwing) [reply project/MimaExcludes.scala]: But the user may write some codes to run different logics according to the InputStream types.

Comment (zsxwing) [reply project/MimaExcludes.scala]: By the way, I'm not sure if we want to pursue strictly compatible. Just to point out the issue here.

**Pull request closed without merging**